### PR TITLE
Rework compose box

### PR DESCRIPTION
**Pull request by satya164**

### CODE REVIEW

File: ui/components/compose.jsx [hunk 1]
```diff
@@ -77,36 +45,72 @@ module.exports = function(core, config, store) {
...
+				msg = input || "",
+				user, nick, atStart;
+
+			if (textObj) {
+				nick = appUtils.formatUserName(textObj.from);
+				user = appUtils.formatUserName(store.get("user"));
 
-			// FIXME: figure out why replaceState is not working
-			composeBox.innerText = composeBox.textContent = this.getInitialState().userInput;
+				if (/^@\S+[\s+{1}]?/.test(msg)) {
```
Comment (aravindet) [inline ui/components/compose.jsx]: What is this regex? Is it intended that mentions terminate in either space, `+`, `{` and `}`?

Perhaps you meant just `/^@\S+\s?/`?

Comment (aravindet) [inline ui/components/compose.jsx]: If I understand your logic correctly, you’re trying to do this:
```
hello -> hello @you
@someone hey -> @you hey
hello @someone -> hey @someone
hello @someone more -> hello @someone more @you
```

May I suggest a simpler option to do the whole thing?
```
msg = msg.replace(/(?:^@\S+\s?)|(?:\s*(?:@\S+)?$)/, function (match, i, str) {
  if (i===0) return "@" + nick + (str !== match? " ": "") // at beginning or empty string
  else if (i + match.length === str.length) return " @" + nick;
  else return "";
});
```

Comment (satya164) [inline ui/components/compose.jsx]: @aravindet `/^@\S+[\s+{1}]?/` is the same as `/^@\S+\s?/`or `/^@\S+\s{1}/` (I'll change it).

Comment (satya164) [inline ui/components/compose.jsx]: @aravindet `/^@\S+[\s+{1}]?/` will match both `@hello` and `@hello ` while `/^@\S+\s?/` only matches `@hello `.

File: ui/components/compose.jsx [hunk 2]
```diff
@@ -115,25 +119,37 @@ module.exports = function(core, config, store) {
...
+			}
+
+			next();
+		},
+
 		componentDidMount: function() {
-			this.setPlaceHolder();
+			this.focusInput();
+
+			core.on("statechange", this.onStateChange, 100);
```
Comment (aravindet) [inline ui/components/compose.jsx]: This is weird. On statechange you are getting re-rendered anyway. You can put the this.onStateChange logic in a componentWillUpdate function.

You also dont need to listen onInput or onPaste... inside componentWillUpdate, you can read the entered text from the DOM before it is overwritten, get the currently selected text from store, run the getMessageText algorithm if needed and call this.setState({userInput}) with result.

File: ui/components/compose.jsx [hunk 3]
```diff
@@ -77,36 +45,72 @@ module.exports = function(core, config, store) {
...
+			}
 		},
 
 		onPaste: function() {
 			setTimeout(function() {
-				var text = this.refs.composeBox.getDOMNode().innerHTML;
+				var text = React.findDOMNode(this.refs.composeBox).innerHTML;
 
 				this.setState({ userInput: text });
 			}.bind(this), 10);
```
Comment (aravindet) [inline ui/components/compose.jsx]: Don't do this :(

Comment (satya164) [inline ui/components/compose.jsx]: What's wrong?

File: ui/components/compose.jsx (hunk 2 above)
Comment (satya164) [inline ui/components/compose.jsx]: On global statechange the component doesn't update as the state of the component remains the same. The component only updates when it's state changes.

If I don't update the state on input, then if "setState" is called somewhere else, the text will get overwritten. I cannot determine if the newState.userInput should replace the entered text or not just by looking at them. Also, not updating state onInput means the state and the component remain in inconsistent state, which can cause further issues.

I need to listen to onPaste as I need to strip the formatting of the pasted html.

Comment (satya164) [inline ui/components/compose.jsx]: On global statechange the component doesn't update as the state of the component remains the same. The component only updates when it's state changes.

If I don't update the state on input, then if "setState" is called somewhere else, the text will get overwritten. I cannot determine if the newState.userInput should replace the entered text or not just by looking at them, also I cannot update the component state from within componentWillUpdate.

Not updating state onInput means the state and the component remain in inconsistent state, which can cause further issues.

I need to listen to onPaste as I need to strip the formatting of the pasted html.

I think it's always better to depend on a single state (either the component state or the global state) to update the component. The component state updates in various situations anyways, so I chose to make the component only update when its internal state changes. Makes the whole thing much simpler and easily to understand.

**Comment by satya164** [comment]
There seem to be a lot of issues with using contenteditable. Lack of browser support for input event, browser bugs etc.

It would be much better to resort to good old textarea, which will solve a lot of our issues. I don't even remember why the decision to use contenteditable was taken!

**Comment by satya164** [comment]
Ah, the auto resize thing! How difficult can it be to add autoresize?

**Comment by satya164** [comment]
http://alistapart.com/article/expanding-text-areas-made-elegant

**Comment by satya164** [comment]
http://codepen.io/satya164/pen/QwoooL

**Pull request closed without merging**